Debug avg planner

Alvinosaur opened the pull request:
Hey Sam, I'm sorry for  leaving all those random txt and deprecated_reference files. I tried to to first make an orphan branch and unstage all files. Then to make sure this branch was comparable to the debug_avg_planner branch, I tried rebasing on top of commit #1, but this just messed things up. Please just ignore those  other files. I've added comments to the relevant files:

main.py
sim_objects.py
environment.py

You can also see that I run experiments by printing out text and redirecting output from stdout to a text file. I also store random things in .npz files.

Alvinosaur committed:
new progress on setting up arm bottle simulation

Alvinosaur committed:
functional arm initial position

Alvinosaur committed:
arm correctly swings and hits cylinder

Alvinosaur committed:
functional video logging, need to install ffmpeg

Alvinosaur committed:
functional restart of simulation

Alvinosaur committed:
gave up on scipy optimize for finding max reach arm

Alvinosaur committed:
running but untested

Alvinosaur committed:
almost fully functional, need to log results and finalize other tests

Alvinosaur committed:
arm colliding correctly with bottle, doesn't slide though

Alvinosaur committed:
functional distribution plotting and switching between viz/non-viz simulation

Alvinosaur committed:
all tests finished

Alvinosaur committed:
fully functioning tests, cleaned up code

Alvinosaur committed:
fixed tests

Alvinosaur committed:
bug with changing center of mass

Alvinosaur committed:
correct change center of mass, some modified tests: only consider distance if not fallen

Alvinosaur committed:
make arm velocity tolerance-testing, add bottle shape test, basic bottle grabbing

Alvinosaur committed:
add basic mdp for understanding

Alvinosaur committed:
rename

Alvinosaur committed:
functional testing actions as direction vectors

Alvinosaur committed:
functional pushing and enviornment setup class

Alvinosaur committed:
fully functional pushing

Alvinosaur committed:
functional modular simulation env, need to finish mdp

Alvinosaur committed:
functional full mdp for diff  x,y

Alvinosaur committed:
made temporary changes to action space for testing, changed center of mass applying but doesn't fix issue

Alvinosaur committed:
value iteration seems kind-of working, simulation too slow

Alvinosaur committed:
move old files to separate folder

Alvinosaur committed:
saving old code

Alvinosaur committed:
save old results

Alvinosaur committed:
clean up

Alvinosaur committed:
issue with arm crashing after solving IK

Alvinosaur committed:
functional initial planning!

Alvinosaur committed:
not sure  about basic cayley quaternion discretization

Alvinosaur committed:
functional integration of bottle orientation into state space

Alvinosaur committed:
isolate  planner  from execution to allow for future interleaved replanning and execution
NONfunctinoal, there was a bug with arm resetting

Alvinosaur committed:
clean up, move bottle ori from state to transition metadata:
add draft of running sim and returning mode of results

Alvinosaur committed:
add invalid transitions and have envrionment return joint pose directly

Alvinosaur committed:
heuristic based on nearest joint position to bottle works

Alvinosaur committed:
direction-guidance looks reasonable

Alvinosaur committed:
functional planning

Alvinosaur committed:
functional replaying of planned trajectory

Alvinosaur committed:
plans are too simple

Alvinosaur committed:
generated a set of reasonble start, goal pairs with enough diversity

Alvinosaur committed:
almost works for complex plans, but limited by IK solver for joints from EE pos

Alvinosaur committed:
make random config generation at joint-level, not EE position

Alvinosaur committed:
temp

Alvinosaur committed:
found filtered set of feasible start, goal pairs

Alvinosaur committed:
found basic plan  for  all objects

Alvinosaur committed:
temp

Alvinosaur committed:
demonstrated problem/purpose: single planner can knock bottle over during execution

Alvinosaur committed:
temp

Alvinosaur committed:
demonstrated use-case with average planner succeeding

Alvinosaur committed:
temp, debugging why avg plannner fails

Alvinosaur committed:
plotting joint traj comparsion

Alvinosaur committed:
fixed discrepancy btwn planned traj and executed traj: use state path, not policy

Alvinosaur committed:
new plan results

Alvinosaur committed:
temp

Alvinosaur committed:
verified state-path-based execution is better to use than direct actions; extended simulation time to address bottle falling

SamNPowers commented on sim_objects.py:
```diff
@@ -0,0 +1,277 @@
+        return Bottle.PLASTIC_MASS + (
+            fill_prop * self.max_volume * Bottle.VOL_TO_MASS)
+
+    def com_from_fill(self, fill_prop):
```
[Minor] "com" -> "center_of_mass" would be clearer for new people reading your code

Alvinosaur replied:
Makes sense

SamNPowers commented on sim_objects.py:
```diff
@@ -0,0 +1,277 @@
+        #     p.removeBody(self.bottle_id)
+        if ori is None:
+            ori = [0, 0, 0, 1]
+        if pos is not None:
```
Unless I'm misreading, you can just do:
```
if pos is None:
    pos = self.start_pos
```

and then you don't need the current if/else (just use pos)

Alvinosaur replied:
Oh wow yes!

SamNPowers commented on sim_objects.py:
```diff
@@ -0,0 +1,277 @@
+                 fric=DEFAULT_FRIC,
+                 radius=DEFAULT_RAD,
+                 height=DEFAULT_HEIGHT):
+        self.start_pos = start_pos
```
Not sure yet if this applies, but just so you know it's common (and convenient) for private variables to be "\_my\_var" and publicly accessible ones to be "my\_var". I.e. the difference being the starting "\_". (Perhaps you know this and all of these are public.) 

This is convenient for people using your code, to know what to expect - I usually make stuff private by default, unless I have made an explicit decision to make it acceptable for others (other consumers of the class) to use.

SamNPowers replied:
By the way this applies to methods too, not just class variables.

Alvinosaur replied:
Ah yes, I am aware of this convention, but was just lazy about it. I will make this change later since it affects many more things, but it definitely would help me avoid making hacky fixes that directly overwrite the private variables, a sign that I should write a public function to do this.

SamNPowers commented on sim_objects.py:
```diff
@@ -0,0 +1,277 @@
+    def __init__(self, EE_start_pos, start_ori, kukaId):
+        self.EE_start_pos = EE_start_pos
+        self.start_ori = start_ori
+        self.kukaId = kukaId
```
[Very Minor] Inconsistent with your snake case everywhere else. (Not gonna make this comment everywhere, but it applies to everything camel case other than class names.)

SamNPowers commented on sim_objects.py:
```diff
@@ -0,0 +1,277 @@
+        self.kukaId = kukaId
+        self.base_pos = np.array([0, 0, 0.1])
+        self.min_dist = 0.3
+        self.MAX_HORIZ_REACH = 0.851  # need to set with set_general_max_reach()
```
set_general_max_reach() is setting "MAX_REACH" not "MAX_HORIZ_REACH", so not sure if the code is doing something other than what you intended.

By the way, if I *need* something to be set, I'll usually use the @property decorator. I.e. I'll do something along the lines of

```
def __init__(self):
    self._some_var = None

@property
def some_var():
    assert self._some_var is not None, "some_var must be set before it can be used."
    return self._some_var
```

And then set _some_var however you like. (TIL there's a fancy @some_var.setter decorator that can be used, though set_general_max_reach() is fine too)

Alvinosaur replied:
That is a great idea, having the get_general_max_reach would also be good, and both could make use of the example syntax:

temperature = property(get_temperature, set_temperature)

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+    # env.bottle = new_bottle
+
+    # run planner and visualize result
+    num_sims_per_action = 20
```
I find it convenient to centralize all my constants, so when I want to change one I don't have to go hunting around for it. I know in academia stuff is more ad-hoc so not really stating this as a problem.

But I do like the pattern you've established of naming the variable before using it, so you're clear on what the number is supposed to mean (vs just using it in-place).

Alvinosaur replied:
Oh yes, I will move all these constants to the top of the script so I know where to look next time. A lot of the bugs I saw in the past did involve adding hacky changes to these constants halfway in the code. I would see the initial definition of a constant, but the constant would unexpectedly change later on. By forcing everything to be defined at the top, this bug won't happen.

SamNPowers reviewed (commented):
Just spent some time familiarizing myself with the broad strokes of the code, and making some comments as I noticed things I wanted to comment on. I'm not finished, really just starting, but apparently github will lose my "pending" changes if, for instance, my computer runs out of application memory and decides to close my browser (which happens to me not infrequently at the moment), so submitting an unfinished review. 

I'll keep reviewing over the next few days.

Alvinosaur commented:
> Just spent some time familiarizing myself with the broad strokes of the code, and making some comments as I noticed things I wanted to comment on. I'm not finished, really just starting, but apparently github will lose my "pending" changes if, for instance, my computer runs out of application memory and decides to close my browser (which happens to me not infrequently at the moment), so submitting an unfinished review.
> 
> I'll keep reviewing over the next few days.

Great, thanks Sam! Please take your time, no rush. After the quiz I'll start addressing the comments as well.

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+import pybullet as p
```
One thing I'll say here is that code review is partially for your future self (make sure the architecture is good, stuff is tested, etc) but also it's particularly useful such that *other people* can look at and work with your code. So some of my comments will be with that in mind. If you don't care about that use-case, feel free to ignore stuff in that vein. 

Relatedly: It's convenient to have a requirements.txt file so people can pip install all your dependencies at once. Alternatively if you want (others) to be able to pip install your project, you can have a setup.py file. Or a conda yml file to enable conda. Whatever, just some way for people to be able to quickly setup and get started with your project. (It's also possible to have all three from just one list of dependencies, to enable people to use whatever they like without having to maintain several files containing requirements.)

Alvinosaur replied:
Oh definitely, for this particular project I didn't expect anyone else to use it which is why I didn't do extensive testing, but I honestly should have just done this since it would fix many bugs. These comments are great and I should just follow these best practices regardless of whether people will use the code. 

Yea a requirements file would be good.

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+
+    # exec_params_set = plan_params_sets[0]
+    single_planner.sim_params_set = plan_params_sets
+    avg_planner.sim_params_set = plan_params_sets[:10]
```
As-is it seems odd to set these like this, rather than passing into plan() or passing into constructor (not 100% sure yet what these params are used for). Looking further it looks like there's a change_param_set function that should be getting used?

Also it's odd to get these from the environment, pass it into the planner, then have the planner pass them right back into the environment in run_sim? Why the indirection of parameters, rather than just letting environment be the only owners of them? (Or perhaps I'm missing something?)

Alvinosaur replied:
Yes you are completely correct. I wanted the environment to have be responsible for generating environment physics parameters, but I also wanted to give the planner flexibility in choosing which physics parameters to use during planning in case I wanted to do some experiments. In hindsight, I think I should just make the environment solely responsible though and just call some environment function to do this.

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+        #     plots[r][c].set_title("Joint %d" % i)
+
+        # plots[1][2].legend()
+        # plt.show()
```
I normally say to remove dead code. If you're actively testing stuff out with dead code, that's one thing, but if it sticks around forever it just gets in the way and makes everything harder to read. Plus git keeps everything so it's not like you can't get it back if you want it.

Alvinosaur replied:
Agreed

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+        # plt.show()
+
+
+def main():
```
I would consider breaking this method up into something like:
env = construct_environment()
plan_params = get_plan_params()
planner = create_planner(plan_params)  # Could roll get_plan_params into this, instead
plan = planner.plan()
execute_plan(plan)  # Or perhaps test_plan? It looks like that's more aligned with what the play_results branch above is doing.

You have all the parts here already, it would just be clearer what's happening if it were broken up. Plus then it would be a bit easier to drop in, say, variations of different environments by having different construct_environments() that you are selecting from. Also you could unit test, eg, your execute_plan with some simple test plan in some simple env directly, without first having to use your planner to make it.

SamNPowers replied:
Looking a bit more, not sure why there's so much environment construction stuff not in the environment definition itself?

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+        bpos=startb, arm=arm, joints=start_joints)
+    goal_state = helpers.bottle_EE_to_state(bpos=goalb, arm=arm)
+    planner.start = start_state
+    planner.goal = goal_state
```
You're passing a start and goal into the planner already. Why set them manually like this? Rather than doing this initialization above and passing the right start/goal in from the beginning?

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+    cur_joints_state = np.copy(init_joints)
+    piecewise_trajs_policy = []
+    piecewise_trajs_state = []
+    for i, (dq_vec, num_iters) in enumerate(policy):
```
"dq" -> words?

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+        if use_policy:
+            full_arm_traj = np.vstack(piecewise_trajs_policy)
+        else:
+            full_arm_traj = np.vstack(piecewise_trajs_state)
```
full_arm_traj doesn't seem to be getting used?

SamNPowers commented on main.py:
```diff
@@ -0,0 +1,313 @@
+        else:
+            full_arm_traj = np.vstack(piecewise_trajs_state)
+    except Exception as e:
+        print("Couldn't stack piecewise trajectories b/c %s" % e)
```
You're only really checking one of the stacks, but you then return both from the function? If this stacking is a risk, it seems like you want to do both in the try/except and return the created vals.

SamNPowers commented on naive_joint_space_planner.py:
```diff
@@ -0,0 +1,469 @@
+
+        # method of simulating an action
+        self.sim_mode = sim_mode
+        if sim_mode == self.SINGLE:
```
As-is it doesn't really look like this mode is necessary? Can just use AVG with num_rand_samples=1 and sim_param_set of size 1. Using consistent code paths is generally preferable.

SamNPowers commented on naive_joint_space_planner.py:
```diff
@@ -0,0 +1,469 @@
+        self.da = da_rad
+        self.num_joints = env.arm.num_joints
+        self.A = ActionSpace(num_DOF=self.num_joints, da_rad=self.da)
+        self.G = dict()
```
I guess I'll stop commenting on this, but short variable names are never my preference, particularly 1 or 2 letter ones. What is G? I can only guess.

SamNPowers commented on naive_joint_space_planner.py:
```diff
@@ -0,0 +1,469 @@
+            # 1 is ok, can change, but to double-check that we're not using default value
+            assert(self.num_rand_samples > 1)
+        else:
+            print("Invalid sim mode specified: {}, defaulting to SINGLE".format(sim_mode))
```
Generally preferable to just bail out entirely if there's something as unexpected as this, rather than quietly covering for the user. Preferable to fail fast in unexpected conditions than let bad experiments continue.

SamNPowers commented on naive_joint_space_planner.py:
```diff
@@ -0,0 +1,469 @@
+                 nearest_arm_pos_i=nn_joint_i,
+                 nearest_arm_pos=nn_joint_pos)]
+        closed_set = set()
+        self.G = dict()
```
Not really sure why this is a class variable? Perhaps I'm missing it, but it looks like it's really only used in plan(), and if it is used by a consumer of the planner, maybe it would be better to return it from plan rather than store it on the object? More expected pattern, anyway.

SamNPowers commented on naive_joint_space_planner.py:
```diff
@@ -0,0 +1,469 @@
+                new_arm_positions = self.env.arm.get_joint_link_positions(
+                    next_joint_pose)
+                trans_cost = (self.calc_trans_cost(n, new_arm_positions) +
+                              1 * fall_prob)
```
1 * ? (I assume the 1 was at some point something non-1, but regardless it seems like it should be named and defined, possibly in the constructor with your other planner consts?

SamNPowers commented on naive_joint_space_planner.py:
```diff
@@ -0,0 +1,469 @@
+                arm_bottle_dist, nn_joint_i, nn_joint_pos = (
+                    self.dist_arm_to_bottle(next_state, new_arm_positions))
+
+                # Quick FIX: use EE for transition costs so set nn_joint to EE
```
I've seen "EE" a lot but I do not know what it means. End effector? Preferable to be clear about it, for others reading.

SamNPowers reviewed (commented):
Still not done, maybe about half. Should finish the rest tonight or maybe tomorrow.

Alvinosaur committed:
prove that simulator is not fully deterministic

Alvinosaur committed:
planner execution is deterministic and matches plan!

SamNPowers commented on environment.py:
```diff
@@ -0,0 +1,422 @@
+        return is_fallen, is_collision, self.bottle.pos, self.bottle.ori, self.arm.joint_pose
+
+    def simulate_plan_online(self, init_joints, policy, bottle_pos, bottle_ori, sim_params: EnvParams):
+        """Run simulation with given joint-space trajectory. Does not reset arm
```
simulate_plan and simulate_plan_online seem to have the same doc string, though they do not appear to be doing quite the same thing - mostly seems like a difference of policy specification? (delta joints that get linspace-smoothed, vs just straight commanding the targets?) If I'm right about the difference, not sure "online" is the word I would use to convey that difference? (also the args/returns are out of date)

SamNPowers commented on environment.py:
```diff
@@ -0,0 +1,422 @@
+        return s
+
+
+def test_environment_avg_quat():
```
This is begging to be put into a unit test.

SamNPowers commented on environment.py:
```diff
@@ -0,0 +1,422 @@
+            self.bottle_fill, self.bottle_fric, self.bottle_fill_prob, self.bottle_fric_prob)
+
+
+class Environment(object):
```
Consider if you were going to release this environment so that others could use it (to write their own methods to compare to yours, for instance), what would the easiest API be for them to use? I personally quite like the OpenAI gym spec, but whatever, as long as it's usable. (Where my definition of "environment" here also generally includes reward definition, i.e. target specification.) In your experiment_helpers you have a function that looks like it's generating a goal? AFAICT that should maybe be in Environment. And you have some object definitions in main.py - AFAICT that should also be in environment. And maybe have Environment only have a step function (and reset etc), and a different SimulationRunner or something class that knows how to pass trajectories of actions into the Environment. 

So basically Environment is responsible for its creation (objects and target) and whatever the "core" functionalities are that the env needs. And then for stuff that you expect others to be less interested in using (your trajectory running code, perhaps) you have a separate object. That also has the advantage of you can have multiple different types of trajectory runners, since it looks like you have several subtypes going on within this code (run_sim vs run_multiple_sims, or "online" vs not-online, or whatever makes the most sense).

SamNPowers commented on joint_space_mdp.py:
```diff
@@ -0,0 +1,482 @@
+        plt.legend([str(action) for action in self.A], loc='upper left')
+        plt.show()
+
+    def test_reach_p(self):
```
More good places for unit tests! (Btw if you want advice on how best to do unit tests in python, let me know.)

SamNPowers commented on joint_space_mdp.py:
```diff
@@ -0,0 +1,482 @@
+        self.V = dict()
+        self.P = dict()
+
+    def state_to_key(self, state):
```
"key"? Quite general

SamNPowers commented on joint_space_mdp.py:
```diff
@@ -0,0 +1,482 @@
+                        action.height):
+                    continue
+
+                value, ns = self.env.run_sim_stochastic(action)
```
Where is run_sim_stochastic defined? It looks like it's expected on Environment, but doesn't seem to be there?

SamNPowers commented on joint_space_mdp.py:
```diff
@@ -0,0 +1,482 @@
+    def init_action_space(self, run_full_mdp):
+        A = []  # need to maintain order
+        self.da = math.pi/80
+        if run_full_mdp:
```
Not sure why this boolean would map to what's happening in the body of the if/then. Comment? (Btw I normally think of variable names and code logic as "what" and comments as "why" - so if the why is clear, no comments necessary, but in this case I find the why not very clear.)

SamNPowers reviewed (commented):
I'm so sorry this took me so long to look through all of (rebuttal involved more work than I thought). 

Overall I found it quite readable, and generally, I think, understandable.

Here's what my biggest piece of feedback is: consider separating things out as if you plan on re-using them (or plan on releasing it for others to use). 

If someone was going to develop their own method using this bottle-arm environment, how easy would that be for them? (Would they have to do a whole bunch of target/bottle/arm initialization on their own?)

If you want to apply your planner to something that isn't this bottle-arm environment, how easy would *that* be? "Bottle" and "arm" specific stuff appear all over the place. Perhaps you create a BottleArmEnvironment, then maybe a Planner class with a BottleArmPlanner subclass, where generic stuff goes in Planner, and then specific stuff goes in the subclass (e.g. compute_heuristic) or whatever. Then when you want to apply your planning logic to some other environment all the shared stuff would be re-usable, and you'd only have to specify the bare minimum necessary. It seems like NaivePlanner is aaalmost there in terms of this, but then a bunch of arm-bottle specific stuff occurs in lots of places, e.g. in main.py. Consolidate it so it's easier to change out, and your future self will thank you.

Unit tests help with having clear APIs. Define what you want your shared, re-usable interface to be, then write a unit test where everything except what you're testing is spoofed. E.g. write a fake Environment that your NaivePlanner can use. That'll quickly give you a good sense for how hard/easy it'll be to re-use your code on a different task. And vice versa, create a simple "policy" that uses the environment (random, perhaps) - how easy is it to set up the Environment?

Let me know if you have any questions, want to hop on a call, etc.

Alvinosaur committed:
deterministic planning, fix bottle jumping issue

Alvinosaur committed:
demonstrated why and how avg planner can be better and worse

Alvinosaur committed:
temp

Alvinosaur committed:
clean up part 1

Alvinosaur committed:
initial experiments with different thresholds and parameter sampling strategies

Alvinosaur committed:
clean up experiment setup

Alvinosaur committed:
functional training pipelines, next will explore use of parent state to simulate action

Alvinosaur committed:
experiment with friction and fill, verify correct distributions

Alvinosaur committed:
remove temp.py

Alvinosaur committed:
clean up pipeline

Alvinosaur committed:
temp

Alvinosaur committed:
update start-goals to be feasible but not trivially easy, clean up args

Alvinosaur committed:
promising results showing motivation, but high friction seems unrealistic

Alvinosaur committed:
start-goals 0 - 11 inclusive all can be planned with bimodal

Alvinosaur committed:
stable commit before debugging low success rates

Alvinosaur committed:
properly initialize bottle orientation when starting planning from a existing planned path

Improve detection of "No contact" simulations to quickly terminate simulation

Include phi (bottle vertical tilt) in cost computation to discourage tipping the bottle

Include bottle position and phi variance in cost computation to explicitly include uncertainty in cost

ADjust distance cost weights when computing heuristic to increase weight for height z so arm moves more towards mid height of bottle

Increase goal threshold to be 0.2 since more realistic and feasible

Increase weighted epsilon from 5 to 7 to account for higher weighted costs

Alvinosaur committed:
debugging RTDP

Alvinosaur committed:
draft of lazy evaluation and reinsertion

Alvinosaur committed:
got rid of nn joint heuristic temporarily

Alvinosaur committed:
add timing benchmarks

Alvinosaur committed:
remove resets between state-action transitions

Alvinosaur committed:
fix lazy search implementation

Alvinosaur committed:
minor changes

Alvinosaur committed:
saving working setup

Alvinosaur committed:
commit before beta distribution work

Alvinosaur committed:
temporary refactoring of code

Alvinosaur committed:
final open-loop, refactoring, improved auto-generation of start-goals

Alvinosaur committed:
functional refactoring of code, use beta-guided planning, 30 new pairs of start-goals complete auto-generated by lazy planner

Alvinosaur committed:
not so great performance, but clean parsing/analysis of results
